go/vt/vtgate/vindexes/cfc.go

```diff
@@ -0,0 +1,326 @@
+
+		if _, err := buf.Write(vind.hash(v[prev:offset])); err != nil {
+			return nil, err
+		}
+		if n == offset {
+			break
+		}
+		prev = offset
+	}
+	if offset < n {
+		if _, err := buf.Write(vind.hash(v[offset:n])); err != nil {
+			return nil, err
+		}
+	}
+	return buf.Bytes(), nil
+}
+
+// Verify returns true if ids maps to ksids.
+func (vind *CFC) Verify(_ VCursor, ids []sqltypes.Value, ksids [][]byte) ([]bool, error) {
+	out := make([]bool, len(ids))
+	for i := range ids {
+		v, err := vind.computeKsid(ids[i].ToBytes(), false)
+		if err != nil {
+			return nil, err
+		}
+		out[i] = bytes.Equal(v, ksids[i])
+	}
+	return out, nil
+}
+
+// Map can map ids to key.Destination objects.
+func (vind *CFC) Map(cursor VCursor, ids []sqltypes.Value) ([]key.Destination, error) {
+	out := make([]key.Destination, len(ids))
+	for i, id := range ids {
+		v, err := vind.computeKsid(id.ToBytes(), false)
+		if err != nil {
+			return nil, err
+		}
+		out[i] = key.DestinationKeyspaceID(v)
+	}
+	return out, nil
+}
+
+// PrefixVindex switches the vindex to prefix mode
+func (vind *CFC) PrefixVindex() SingleColumn {
+	return vind.prefixVindex
+}
+
+func addOne(value []byte) []byte {
+	n := len(value)
+	overflow := true
+	for i := n - 1; i >= 0; i-- {
+		if value[i] < 255 {
+			value[i]++
+			overflow = false
+			break
+		} else {
+			value[i] = 0
+		}
+	}
+	if overflow {
+		return nil
+	}
+	return value
+}
+
+type prefixCFC struct {
+	name string
+	cfc  *CFC
+}
+
+func (vind *prefixCFC) String() string {
+	return vind.name
+}
+
+func (vind *prefixCFC) Cost() int {
+	return 1
+}
+
+func (vind *prefixCFC) IsUnique() bool {
+	return false
+}
+
+func (vind *prefixCFC) NeedsVCursor() bool {
+	return false
+}
+
+// Verify returns true if ids maps to ksids.
+func (vind *prefixCFC) Verify(vc VCursor, ids []sqltypes.Value, ksids [][]byte) ([]bool, error) {
+	// prefixCFC is only active for 'LIKE' expr. Verify is used by 'INSERT' so we re-use CFC.Verify()
+	return vind.cfc.Verify(vc, ids, ksids)
+}
+
+// Map can map ids to key.Destination objects.
+func (vind *prefixCFC) Map(cursor VCursor, ids []sqltypes.Value) ([]key.Destination, error) {
+	out := make([]key.Destination, len(ids))
+	for i, id := range ids {
+		value := id.ToBytes()
+		// we only consider the string up to the first whild char as the prefix.
+		p := bytes.IndexByte(value, '%')
+		if p < 0 {
+			// if there is no wild char, we use the whole string
+			p = len(value)
+		}
+		begin, err := vind.cfc.computeKsid(value[:p], true)
+		if err != nil {
+			return nil, err
+		} else if len(begin) == 0 {
+			// prefix is too short, this is typically the case when the prefix
+			// is smaller than the first offset when using hash.
+			out[i] = key.DestinationAllShards{}
+		} else {
+			// the prefix maps to keyspace id corresponding to value and plus one.
+			// that is [ keyspace_id, keyspace_id + 1 ).
+			end := make([]byte, len(begin))
+			copy(end, begin)
+			end = addOne(end)
+			out[i] = key.DestinationKeyRange{
```

Comment:
This entire block might be better suited as a constructor named something like `NewKeyRangeFromPrefix` that lives alongside `topodatapb.KeyRange` or `key.DestinationKeyRange`?

Reply:
Good point. Changed.